R/conditional-transform.R

```diff
@@ -0,0 +1,150 @@
+#' side is where to get the value if the condition on the righthand side is 
+#' true. This is commonly a Crunch variable but may be a string or numeric 
+#' value, depending on the type of variable you're constructing.
+#'
+#' `conditionalTransform` is similar to `makeCaseVariable`; however, 
+#' `conditionalTransform` can use other Crunch variables as a source of a 
+#' variable, whereas, `makeCaseVariable` can only use characters. This 
+#' additional power comes at a cost: `makeCaseVariable` can be executed 
+#' entirely on Crunch servers, so no data needs to be downloaded or uploaded 
+#' to/from the local R session. `conditionalTransform` on the other hand will 
+#' download the data necessary to construct the new variable.
+#' 
+#' For more details see the \code{vignette("conditional-variables", package="crunch")} vignette.
+#'
+#' @param ... a list of cases to evaluate as well as other
+#' properties to pass about the case variable (i.e. alias, description)
+#' @param data a Crunch dataset object to use
+#' @param else_condition a default value to use if none of the conditions are 
+#' true (default: `NA`)
+#' @param type a character that is either "categorical", "text", "numeric" what
+#'  type of output should be returned? The source variables will be converted 
+#'  to this type if necessary
+#' @param categories a vector of characters if `type="categorical"`, these are 
+#' all of the categories that should be in the resulting variable, in the order
+#' they should be in the resulting variable or a set of Crunch categories.
+#'
+#' @return a Crunch `VariableDefinition`
+#' @examples
+#' \dontrun{
+#' 
+#' ds$cat_opinion <- conditionalTransform(pet1 == 'Cat' ~ Opinion1,
+#'                                        pet2 == 'Cat' ~ Opinion2,
+#'                                        pet3 == 'Cat' ~ Opinion3, 
+#'                                        data = ds,
+#'                                        name = "Opinoin of Cats")
+#' }
+#' @export
+conditionalTransform <- function (..., data, else_condition=NA, type="categorical", categories=NULL) {
+    dots <- list(...)
+    is_formula <- function (x) inherits(x, "formula")
+    formulas <- Filter(is_formula, dots) 
+    var_def <- Filter(Negate(is_formula), dots)
+    n <- length(formulas)
+    
+    if (n == 0) {
+        halt("no conditions have been supplied; please supply formulas as conditions.")
+    }
+    if (!type %in% c("categorical", "text", "numeric")){
+        halt("type must be either ", dQuote("categorical"), ", ", dQuote("text"), ", or ", dQuote("numeric"))
+    }
+    if (type != "categorical" & !is.null(categories)){
+        halt("type is not ", dQuote("categorical"), " ignoring ", dQuote("categories"))
+    }
+    
+    cases <- vector("list", n)
+    values <- vector("list", n)
+    for (i in seq_len(n)) {
+        formula <- formulas[[i]]
+        if (!inherits(formula, "formula") || length(formula) != 3) {
```

Comment:
Yup, this was a holdover from a previous way of testing I've removed the `!inherits(formula, "formula") || ` bit. I don't think there's a situation where `length(formula) != 3` but because we are referencing the formula by indices, I want to keep it just in case there is some weird edge case formula we aren't expecting.